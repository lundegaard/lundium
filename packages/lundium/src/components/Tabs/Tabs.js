import PropTypes from 'prop-types';
import { cx, noop } from 'ramda-extension';
import React, { Children, createElement, forwardRef, useImperativeHandle, useState } from 'react';
import { dissoc } from 'ramda';

import Box from '../Box';
import Tab from '../Tab';
import TabBar from '../TabBar';

const omitTabLabel = dissoc('tabLabel');

const Tabs = forwardRef(
	({ className, defaultActiveIndex, children, onTab = noop, ...otherProps }, ref) => {
		const [activeIndex, setActiveIndex] = useState(defaultActiveIndex || 0);
		const tabBarContent = Children.map(children, (child, index) => {
			const handleTabClick = index => event => {
				setActiveIndex(index);
				onTab(index);
				event.preventDefault();
			};

			const { tabLabel, ...childProps } = child.props;

			return (
				<Tab
					{...childProps}
					isActive={activeIndex === index}
					onClick={handleTabClick(index)}
					label={tabLabel}
				/>
			);
		});

		const childrenArray = Children.toArray(children);
		const activeTab = childrenArray[activeIndex];

		useImperativeHandle(ref, () => ({ setActiveIndex }));

		return (
			<Box className={cx('tabs', className)} {...otherProps}>
				<TabBar>{tabBarContent}</TabBar>
				{createElement(activeTab.type, omitTabLabel(activeTab.props))}
			</Box>
		);
	}
);

Tabs.displayName = 'forwardRef(Tabs)';

Tabs.propTypes = {
	children: PropTypes.node,
	className: PropTypes.string,
	defaultActiveIndex: PropTypes.number,
	onTab: PropTypes.func,
};

export default Tabs;
