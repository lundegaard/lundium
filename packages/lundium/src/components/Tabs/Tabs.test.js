import React from 'react';
import { shallow } from 'enzyme';

import { Text } from '../../index';

import Tabs from './Tabs';

describe('<Tabs />', () => {
	const requiredProps = {};

	it('matches snapshot when shallowly rendered', () => {
		const wrapper = shallow(
			<Tabs {...requiredProps}>
				<Text tabLabel="First">First content</Text>
				<Text tabLabel="Second">Second content</Text>
			</Tabs>,
		);
		expect(wrapper).toMatchSnapshot();
	});
});
