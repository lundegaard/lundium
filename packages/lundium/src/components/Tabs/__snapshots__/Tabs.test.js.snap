// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> matches snapshot when shallowly rendered 1`] = `
<forwardRef(Box)
  className="tabs"
>
  <TabBar>
    <Tab
      isActive={true}
      key=".0"
      label="First"
      onClick={[Function]}
    >
      First content
    </Tab>
    <Tab
      isActive={false}
      key=".1"
      label="Second"
      onClick={[Function]}
    >
      Second content
    </Tab>
  </TabBar>
  <forwardRef(Text)>
    First content
  </forwardRef(Text)>
</forwardRef(Box)>
`;
